src/Core/Silk.NET.Core/Native/SilkMarshal.cs

```diff
@@ -112,61 +112,111 @@ public static GlobalMemory StringToMemory
+                NativeStringEncoding.LPWStr => ((input?.Length ?? 0) + 1) * 2,
+                _ => -1
             };
 
-            static unsafe GlobalMemory Utf8ToMemory(string input)
+        /// <summary>
+        /// Marshals the input string per the specified native string encoding, and stores the marshalled bytes into
+        /// the given span
+        /// </summary>
+        /// <param name="input">The string to marshal.</param>
+        /// <param name="span">The span to marshal the string into.</param>
+        /// <param name="encoding">The target native string encoding.</param>
+        /// <returns>The number of bytes copied into the span including the null terminator.</returns>
+        public static unsafe int StringIntoSpan
+        (
+            string? input,
+            Span<byte> span,
+            NativeStringEncoding encoding = NativeStringEncoding.Ansi
+        )
+        {
+            if (input is null)
+            {
+                span[0] = 0;
+                return 1;
+            }
+            
+            switch (encoding)
             {
-                var memory = GlobalMemory.Allocate(Encoding.UTF8.GetMaxByteCount(input.Length) + 1);
-                int convertedBytes;
-                fixed (char* firstChar = input)
+                case NativeStringEncoding.BStr:
                 {
-                    fixed (byte* bytes = memory)
-                    {
-                        convertedBytes = Encoding.UTF8.GetBytes(firstChar, input.Length, bytes, memory.Length - 1);
-                    }
+                    static void ThrowNoBStr() => throw new InvalidOperationException
+                    (
+                        "BSTR encoded strings can only be marshalled into known BSTR memory."
+                    );
+                    
+                    ThrowNoBStr();
+                    return -1;
                 }
+                case NativeStringEncoding.LPStr:
+                case NativeStringEncoding.LPTStr:
+                case NativeStringEncoding.LPUTF8Str:
+                {
+                    int convertedBytes;
 
-                memory[convertedBytes] = 0;
-                return memory;
-            }
-
-            static unsafe GlobalMemory AnsiToMemory(string input)
-            {
-                var memory = GlobalMemory.Allocate((input.Length + 1) * Marshal.SystemMaxDBCSCharSize);
-                int convertedBytes;
+                    fixed (char* firstChar = input)
+                    {
+                        fixed (byte* bytes = span)
+                        {
+                            convertedBytes = Encoding.UTF8.GetBytes(firstChar, input.Length, bytes, span.Length - 1);
+                        }
+                    }
 
-                fixed (char* firstChar = input)
+                    span[convertedBytes] = 0;
+                    return ++convertedBytes;
```

Comment:
Please forward to Utf8.FromUtf16 on >= netcore3.0
Something like
```cs
#if NET5_0
Debug.Assert(Utf8.FromUtf16(input, span, out _, out convertedBytes, true, true) == OperationStatus.Done);
#endif
```
We probably don't care for the returned status, it can only be done, or theoretically, `DestinationTooSmall` but that can only happen if GetMaxSizeOf fails.
(If you don't think this is fit for this PR, do just reference this in a new issue, and I'll get back to it asap, we absolutely need to track this though, perf of the string marshalling is miserable)

Reply:
Opened an issue.